test/spec/api/v4/dataview/formula.spec.js

```diff
@@ -0,0 +1,207 @@
+    if (key === 'data') {
+      return 1234;
+    }
+    if (key === 'nulls') {
+      return 42;
+    }
+  });
+  _.extend(internalModelMock, Backbone.Events);
+
+  return internalModelMock;
+}
+
+function createSourceMock () {
+  return new carto.source.Dataset();
+}
+
+function createEngineMock () {
+  var engine = {
+    name: 'Engine mock',
+    reload: function () {}
+  };
+  spyOn(engine, 'reload');
+
+  return engine;
+}
+
+describe('formula dataview public v4 API', function () {
+  var source = createSourceMock();
+
+  describe('initialization', function () {
+    it('source must be provided', function () {
+      var test = function () {
+        new carto.dataview.Formula(); // eslint-disable-line no-new
+      };
+
+      expect(test).toThrowError(TypeError, 'Source property is required.');
+    });
+
+    it('column must be provided', function () {
+      var test = function () {
+        new carto.dataview.Formula(source); // eslint-disable-line no-new
+      };
+
+      expect(test).toThrowError(TypeError, 'Column property is required.');
+    });
+
+    it('options set to default if not provided', function () {
+      var column = 'population';
+
+      var dataview = new carto.dataview.Formula(source, column);
+
+      expect(dataview._options).toBeDefined();
+      expect(dataview._options.operation).toEqual(carto.operation.COUNT);
+    });
+
+    it('options set to the provided value', function () {
+      var dataview = new carto.dataview.Formula(source, 'population', {
+        operation: carto.operation.AVG
+      });
+
+      expect(dataview._options.operation).toEqual(carto.operation.AVG);
+    });
+
+    it('throw error if no correct operation is provided', function () {
+      var test = function () {
+        new carto.dataview.Formula(source, 'population', { // eslint-disable-line no-new
+          operation: 'exponential'
+        });
+      };
+
+      expect(test).toThrowError(TypeError, 'Operation for formula dataview is not valid. Use carto.operation');
+    });
+  });
+
+  describe('.setOperation', function () {
+    var dataview;
+
+    beforeEach(function () {
+      dataview = new carto.dataview.Formula(source, 'population');
+    });
+
+    it('checks if operation is valid', function () {
+      var test = function () {
+        dataview.setOperation('swordfish');
+      };
+
+      expect(test).toThrowError(TypeError, 'Operation for formula dataview is not valid. Use carto.operation');
+    });
+
+    it('if operation is valid, it assigns it to property, returns this and nothing else if there is no internaModel', function () {
+      var returnedObject = dataview.setOperation(carto.operation.AVG);
+
+      expect(dataview.getOperation()).toEqual(carto.operation.AVG);
+      expect(returnedObject).toBe(dataview);
+    });
+
+    it('sets operation in internal model if exists', function () {
+      var internalModelMock = createInternalModelMock();
+      dataview._internalModel = internalModelMock;
```

Comment:
Perhaps `$getInternalModel` is a bit more "stable" than `_internalModel`. We could `spyOn(dataview, '$getInternalModel').and.returnValue(internalModelMock)`. This will raise an error if `Base.prototype.$getInternalModel` is ever deleted.

Reply:
Another 💡 .Perhaps we could have `Base.prototype.$setInternalModel` and use it both internally and in our tests.

Reply:
In dataviews there is no need of adding `$setInternalModel`. I would avoid it, allowing creating the `_internalModel` only through the `_createInternalModel` function. I would avoid exposing more functions and giving the impressions that `$setInternalModel` is a common operation in dataviews. So far, there's only a use case that is using it in specs (that are under control).

`$getInternalModel` is not used in specs. I mean, we don't read the internal model directly in specs.